Mostly kill "LValue access kind"

slavapestov opened the pull request:
SourceKit still uses it in a couple of places. I need to refactor it to compute it in a separate pass for its own use instead of having the type checker store this inside Exprs.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@akyrtzi The indexer seems to be the only meaningful place where Expr::getLValueAccessKind() is used, in order to determine if a property is read, written or both. Do you think another approach would work here, possibly by walking the AST and computing this flag in the indexer itself? Or maybe it's not needed at all? This has been a source of crashes and spurious asserts so it would be nice to just remove it, since SILGen doesn't really care about it at all.

slavapestov commented:
@DougGregor @jrose-apple @rjmccall Also if you have thoughts on this please share.

akyrtzi commented:
I'd recommend to have an alternative for `determine if a property is read, written or both` before removing this; it is useful info we can do queries for.

jrose-apple commented:
I'm happy to see it go if we're not using it, but I thought SILGen needed this information anyway to know whether a struct property access required writeback or not (`nonmutating`).

jrose-apple commented:
(SourceKit using it totally counts as using it, by the way.)

DougGregor commented:
It's just SourceKit that's actually using it now. I'd much rather SourceKit gather this info structurally instead of baking it into the AST.

nkcsgexi commented on lib/AST/SourceEntityWalker.cpp:
```diff
@@ -255,7 +249,7 @@ std::pair<bool, Expr *> SemaAnnotator::walkToExprPre(Expr *E) {
     return { true, E };
 
-  Optional<AccessKind> OpAccess = getAccessKind(E);
+  Optional<AccessKind> OpAccess;
```
This is meaningful use of accesskind. Internal codebase needs such info.

slavapestov replied:
I'm going to pull the calculation out into a visitor SourceKit can invoke on-demand, instead of doing it in Sema all the time (since SILGen doesn't need it).

DougGregor replied:
WFM!

slavapestov committed:
SILGen: Remove last usage of getLValueAccessKind()

slavapestov committed:
IDE: Use the is<> template instead of TypeBase::getKind()

slavapestov committed:
AST: Move SourceEntityWalker to IDE

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov committed:
IDE: Remove a LValueAccessKind usage

Ok, so behind the scenes SourceEntityWalker is still using
Expr::getLValueAccessKind(). But now that SourceEntityWalker
is really the *only* remaining client, it makes sense to
refactor the calculation of this into a separate pass, which
only SourceKit needs to use.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov committed:
IDE: Fix bug in SourceEntityWalker subscript handling

We weren't passing the lvalue access kind to the subscript.
This was causing a test to fail with the next patch.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test
